skimage/_shared/testing.py

```diff
@@ -52,6 +52,13 @@ def assert_greater(a, b, msg=None):
     assert a > b, message
 
 
+def assert_shape_equal(im1, im2):
+    """Raise an error if the shape and dtype do not match."""
+    if not im1.shape == im2.shape:
```

Comment:
Why not just `assert im1.shape == im2.shape`? Do we really need this function? The gain looks low to me.

Reply:
See also my comments below.

Reply:
We had `assert im1.shape == im2.shape` previously, @JDWarner thought it might be useful to have a function in `_shared`
See https://github.com/scikit-image/scikit-image/pull/4025#discussion_r303270180
If we made this function use an assert statement would that be an improvement? I'm impartial either way.

```
In [7]: def assert_shape_equal(a, b): 
   ...:     message = "%r does not have equal shape with %r" % (a.shape, b.shape) 
   ...:     assert a.shape == b.shape, message 

In [8]: assert_shape_equal(x,y)                                                                                                                                                   
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-8-8237c8a1005b> in <module>
----> 1 assert_shape_equal(x,y)

<ipython-input-7-2933226b4e44> in assert_shape_equal(a, b)
      1 def assert_shape_equal(a, b):
      2     message = "%r does not have equal shape with %r" % (a.shape, b.shape)
----> 3     assert a.shape == b.shape, message
      4 
      5 

AssertionError: (3, 3) does not have equal shape with (4, 4)

```

Reply:
We should not use assert if the check must be done in any situation (including with optimize flags). I would say no.

My opinion on this is to put this function in _shared/utils.py, below `assert_nD`, and rename these functions to do not use the word "assert". Like check_nD, check_shape_equality.

This can be addressed in another PR. These functions are not in the public API, it won't be painful to change this later.